securedrop/journalist.py

```diff
@@ -593,11 +602,24 @@ def reply():
                         [crypto_util.getkey(g.sid), config.JOURNALIST_KEY],
                         output=store.path(g.sid, filename))
     reply = Reply(g.user, g.source, filename)
-    db_session.add(reply)
-    db_session.commit()
 
-    flash("Thanks! Your reply has been stored.", "notification")
-    return redirect(url_for('col', sid=g.sid))
+    try:
+        db_session.add(reply)
+        db_session.commit()
+    except Exception as exc:
+        flash("An unexpected error occurred! Please check the application "
+              "logs or inform your adminstrator.", "error")
+        # We take a cautious approach to logging here because we're dealing
+        # with responses to sources. It's possible the exception message could
+        # contain information we don't want to write to disk.
+        app.logger.error(
+            "Reply from '{}' (id {}) failed: {}!".format(g.user.username,
+                                                         g.user.id,
+                                                         exc.__class__))
+    else:
+        flash("Thanks! Your reply has been stored.", "notification")
+    finally:
+        return redirect(url_for('col', sid=g.sid))
```

Comment:
> Question: is this PR intended to handle the case where the journalist tries to reply to a recently deleted source?

Yes and no. It's intended to handle the race condition where the source is deleted after `journalist.setup_g()` has finished executing in the request thread, but before `db_session.commit()` is called in `journalist.reply()`. I use the base `Exception` class because I also believe it's possible exceptions other than `StaleDataError` made be raised under certain race-condition-related circumstances (for more see https://github.com/freedomofpress/securedrop/issues/1728).

> will 404 if the source was deleted, so the flashed message above will not be displayed, no?

As explained in https://github.com/freedomofpress/securedrop/pull/1720#issuecomment-305307167, the case in which the source is deleted before the request handling starts will be taken care of in a separate PR. There is a rough WIP branch `gracefully-handle-posts-involving-deleted-sources` (note: am now implementing error handling for _all_ types of requests, not just POSTS).